sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+                ) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            let rhs_ffi_in = {
+                FfiBigNumber {
+                    data: rhs.value.as_ptr() as *const _ as u64,
+                    data_len: rhs.value.len() as u64,
+                    is_negative: rhs.is_negative,
+                }
+            };
+            let value_len = std::cmp::max(lhs_ffi_in.data_len, rhs_ffi_in.data_len) + 1;
+            let mut value = Vec::<u8>::with_capacity(value_len as usize);
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: value_len,
+                    is_negative: false,
+                }
+            };
+            unsafe {
+                sol_bignum_div(
+                    &lhs_ffi_in as *const _ as *const u64,
+                    &rhs_ffi_in as *const _ as *const u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    value,
+                }
+            }
+        }
+    }
+
+    /// Returns self squared
+    pub fn sqr(&self) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            let mut bn_done = openssl::bn::BigNum::new().unwrap();
+            let mut ctx = openssl::bn::BigNumContext::new().unwrap();
+            bn_done.sqr(&self.value, &mut ctx).unwrap();
+            Self { value: bn_done }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_sqr(lhs_ffi_in_addr: *const u64, bn_syscall_result: *mut u64) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            let value_len = lhs_ffi_in.data_len * 2;
+            let mut value = Vec::<u8>::with_capacity(value_len as usize);
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: value_len,
+                    is_negative: false,
+                }
+            };
+            unsafe {
+                sol_bignum_sqr(
+                    &lhs_ffi_in as *const _ as *const u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
```

Comment:
These two lines don't need to be in the `unsafe` block

Reply:
@jackcmay `value.set_len(..)` does need to be in `unsafe` block

Reply:
Why not use one of the safe versions like `truncate`?

Reply:
@jackcmay I thought I had tried that and it failed assertions. I will revisit and research again

Reply:
Cool, should have as little in unsafe blocks as possible :-)

Reply:
@jackcmay As I thought:
Prior to syscall I allocate a Vector with capacity as true result `len` may be less. However; `truncate` relies on the `len` being known to work. If I do not do a `set_len()` it will truncate it to 0!!!